Checkpoint recovery

Random-Liu opened the pull request:
Fixes https://github.com/kubernetes-incubator/cri-containerd/issues/120.

Add recovery logic during restart. I've manually tested this, it's working even better than I thought. :)

Based on https://github.com/kubernetes-incubator/cri-containerd/pull/206 and https://github.com/kubernetes-incubator/cri-containerd/pull/179.

@kubernetes-incubator/maintainers-cri-containerd Please review this carefully, especially `pkg/server/restart.go`. This is really important to the production quality.

I'll send another PR to add our own integration test framework, and will add restart test.

abhinandanpb commented on pkg/server/restart.go:
```diff
@@ -0,0 +1,429 @@
+	}
+
+	// Load network namespace.
+	netNS, err := sandboxstore.LoadNetNS(meta.NetNSPath)
```
NetNSPath is not set if its in host network namespace. Should we be checking for the network namespace before loading it ?

Random-Liu replied:
Good catch. We should.

Random-Liu replied:
Done.

abhinandanpb commented on pkg/server/restart.go:
```diff
@@ -0,0 +1,429 @@
+		}
+		id := d.Name()
+		found := false
+		for _, c := range cntrs {
```
can we use a map to track the loaded containers/sandboxes from the caller. Just thinking,  looping over containers for every directory may not be a faster method.

abhinandanpb replied:
or may be keep a track of containerDirs in a map once loadContainer succeeds. So we can just loop around the dirs here and check if its in the containerDirs map , if not then clean it up.

Random-Liu replied:
Will do.

Random-Liu replied:
Done.

abhinandanpb commented on pkg/server/restart.go:
```diff
@@ -0,0 +1,429 @@
+		}
+		id := d.Name()
+		found := false
+		for _, c := range cntrs {
```
same as mentioned for sandboxes

Random-Liu replied:
Done.

Random-Liu commented:
@abhinandanpb Addressed comments.

abhinandanpb commented on pkg/server/restart.go:
```diff
@@ -0,0 +1,426 @@
+		return container, fmt.Errorf("failed to load task: %v", err)
+	}
+	var s containerd.Status
+	notFound := false
```
var notFound bool

Random-Liu replied:
Done.

abhinandanpb commented on pkg/server/restart.go:
```diff
@@ -0,0 +1,426 @@
+	if err != nil {
+		return fmt.Errorf("failed to list sandbox containers: %v", err)
+	}
+	for _, sandbox := range sandboxes {
```
I think you can create a sandboxmap[string]bool here and pass it to cleanuporphansandboxdir instead of creating one there.

Random-Liu replied:
Doesn't make too much difference to me, because we won't use the map here. Given so I prefer move the code into cleanuporphansandboxdir which is the place where the map will be used. :)

abhinandanpb replied:
I meant instead of passing sandbox list and container list , pass a map for comparison. Not a blocker though

abhinandanpb commented on pkg/server/restart.go:
```diff
@@ -0,0 +1,426 @@
+	if err != nil {
+		return fmt.Errorf("failed to list containers: %v", err)
+	}
+	for _, container := range containers {
```
same as sandbox

Random-Liu commented on pkg/store/sandbox/netns.go:
```diff
@@ -41,6 +50,19 @@ func NewNetNS() (*NetNS, error) {
 }
 
+// LoadNetNS loads existing network namespace. It returns ErrClosedNetNS
+// is the network namespace has already been closed.
```
Done

Random-Liu commented on pkg/store/sandbox/netns.go:
```diff
@@ -49,10 +71,39 @@ func (n *NetNS) Remove() error {
+			return fmt.Errorf("failed to stat netns: %v", err)
+		}
+		// Follow possible /var/run -> /run symlink.
+		path, err := symlink.FollowSymlinkInScope(path, "/")
```
Copied from cri-o. Yeah, it seems that we don't need this even `/var/run -> /run` is a symlink.

abhinandanpb commented on pkg/store/sandbox/netns.go:
```diff
@@ -41,6 +50,19 @@ func NewNetNS() (*NetNS, error) {
+// LoadNetNS loads existing network namespace. It returns ErrClosedNetNS
+// is the network namespace has already been closed.
+func LoadNetNS(path string) (*NetNS, error) {
+	if err := cnins.IsNSorErr(path); err != nil {
```
GetNS would return ```cnins.NSPathNotExistErr``` so you can always check for that error if the nspath doesnt exist. Unless I am missing something. This is fine. If we can avoid an if condition even better.

Random-Liu commented:
> GetNS would return cnins.NSPathNotExistErr so you can always check for that error if the nspath doesnt exist. Unless I am missing something. This is fine. If we can avoid an if condition even better.

@abhinandanpb I read their code. You are right. We should use `cnins.NSPathNotExistErr` instead. We may not want to ignore the error if it's some other transient stat error.

Random-Liu commented:
@abhinandanpb Addressed comments. I still keeps the `ErrClosedNetNS`, so that caller of `LoadNetNS` doesn't need to make assumption on the underlying implementation. However, I does change `LoadNetNS` to look at `cnins.NSPathNotExistErr`.

Random-Liu commented:
@abhinandanpb Thanks for reviewing! Will merge after test passes.

mikebrow commented:
/LGTM
